### PR TITLE
Support executing npm commands

### PR DESCRIPTION
**Pull request by sarao1310**
- [x] All [tests](https://github.com/jfrog/build-info-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] All [static analysis checks](https://github.com/jfrog/build-info-go/actions/workflows/analysis.yml) passed.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

### CODE REVIEW

File: build/npm.go [hunk 1]
```diff
@@ -78,3 +92,13 @@ func (nm *NpmModule) AddArtifacts(artifacts ...entities.Artifact) error {
 	partial := &entities.Partial{ModuleId: nm.name, ModuleType: entities.Npm, Artifacts: artifacts}
 	return nm.containingBuild.SavePartialBuildInfo(partial)
 }
+
+// This function discard the npm command in npmArgs, and keep over only the command flags.
+// It is necessary for the npm command to come before the npm command flags in npmArgs in order to work the function correctly.
```
Comment (eyalbe4) [inline build/npm.go:97]: The above description includes multiple grammar issuues.

File: build/npm.go [hunk 2]
```diff
@@ -78,3 +92,13 @@ func (nm *NpmModule) AddArtifacts(artifacts ...entities.Artifact) error {
...
+func (nm *NpmModule) filterNpmArgsFlags() {
```
Comment (eyalbe4) [inline build/npm.go:98]: Let's add a unit test for this function.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
Let's also update the README accordingly.
I'd be happy to review this PR again following your changes.

File: build/npm.go (hunk 2 above)
Comment (sarao1310) [reply build/npm.go:98]: Done

File: build/npm.go (hunk 1 above)
Comment (sarao1310) [reply build/npm.go]: Done

File: build/utils/npm.go [hunk 3]
```diff
@@ -388,13 +371,21 @@ func RunNpmCmd(executablePath, srcPath string, npmCmd NpmCmd, npmArgs []string,
 	errResult = errBuffer.Bytes()
 	stdResult = outBuffer.Bytes()
 	if err != nil {
-		err = errors.New("error while running the command :'" + executablePath + " " + strings.Join(cmdArgs, " ") + "'\nError output is:\n" + string(errResult) + "\nCommand error: is:\n" + string(err.Error()))
+		err = errors.New("error while running the command :'" + executablePath + " " + strings.Join(tmpArgs, " ") + "'\nError output is:\n" + string(errResult) + "\nCommand error: is:\n" + string(err.Error()))
 		return
 	}
-	log.Debug("npm " + npmCmd.String() + " standard output is:\n" + string(stdResult))
+	// The npm command is the first cell in tmpArgs array.
```
[the remark this answers is not in the document]
Comment (sarao1310) [reply build/utils/npm.go:377]: yes, I changed it.

File: build/utils/npm.go [hunk 4]
```diff
@@ -157,17 +157,17 @@ func installPackageLock(executablePath, srcPath string, npmArgs []string, log ut
 	if npmVersion.AtLeast("6.0.0") {
 		npmArgs = append(npmArgs, "--package-lock-only")
 		// Installing package-lock to generate the dependencies map.
-		_, errData, err := RunNpmCmd(executablePath, srcPath, Install, npmArgs, log)
-		if err != nil {
-			return errors.New("Some errors occurred while installing package-lock: " + string(errData))
+		_, errData, err := RunNpmCmd(executablePath, srcPath, AppendNpmCommand(npmArgs, "install"), log)
+		if err != nil || len(errData) > 0 {
+			return errors.New("Some errors occurred while installing package-lock: " + err.Error())
```
[the remark this answers is not in the document]
Comment (sarao1310) [reply build/utils/npm.go]: Done

File: build/utils/npm.go [hunk 5]
```diff
@@ -388,13 +371,21 @@ func RunNpmCmd(executablePath, srcPath string, npmCmd NpmCmd, npmArgs []string,
...
+		err = errors.New("error while running the command :'" + executablePath + " " + strings.Join(tmpArgs, " ") + "'\nError output is:\n" + string(errResult) + "\nCommand error: is:\n" + string(err.Error()))
 		return
 	}
-	log.Debug("npm " + npmCmd.String() + " standard output is:\n" + string(stdResult))
+	// The npm command is the first cell in tmpArgs array.
+	log.Debug("npm " + tmpArgs[0] + " standard output is:\n" + string(stdResult))
 	return
 }
 
+// This function append the Npm command in the first cell in npmArgs strings array.
```
[the remark this answers is not in the document]
Comment (sarao1310) [reply build/utils/npm.go:382]: Done

File: build/npm_test.go [hunk 6]
```diff
@@ -52,3 +52,37 @@ func TestGenerateBuildInfoForNpm(t *testing.T) {
...
+	// Set arguments in npmArgs.
+	npmArgs := []string{"ls", "--package-lock-only"}
+	_, _, err = buildutils.RunNpmCmd("npm", projectPath, buildutils.AppendNpmCommand(npmArgs, "install"), logger)
+	assert.NoError(t, err)
+	npmModule, err := npmBuild.AddNpmModule(projectPath)
+	assert.NoError(t, err)
+	npmModule.SetNpmArgs(npmArgs)
+	npmModule.filterNpmArgsFlags()
+	var expected []string
+	expected = append(expected, "--package-lock-only")
```
Comment (eyalbe4) [inline build/npm_test.go:81]: The above two lines can be reduced to -
```
expected := []string{"--package-lock-only"}
```

File: cli/cli.go [hunk 7]
```diff
@@ -140,11 +140,16 @@ func GetCommands(logger utils.Log) []*clitool.Command {
...
+				formatValue, filteredArgs, err := extractStringFlag(context.Args().Slice(), formatFlag)
 				if err != nil {
 					return
 				}
-				return printBuild(bld, context.String(formatFlag))
+				npmModule.SetNpmArgs(filteredArgs)
+				err = npmModule.Build()
+				if err != nil {
+					return
+				}
```
Comment (eyalbe4) [inline cli/cli.go:151]: The above line can be reduced to - 
```
				if err = npmModule.Build(); err != nil {
					return
				}
```
Let's implement this change for other commands in this file as well.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
* Some tests are failing.
* See my inline comments.
Feel free to merge after fixing the above.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
Looks like you ignored this comment - 
image

Here as well - 
image

File: build/npm_test.go (hunk 6 above)
Comment (sarao1310) [reply build/npm_test.go:81]: Done!

**Review by eyalbe4 — APPROVED** [review]

**Pull request merged**